proxy/logging/LogObject.cc

```diff
@@ -477,6 +481,89 @@ LogObject::log(LogAccess *lad, const char *text_entry)
   return this->log(lad, std::string_view{text_entry ? text_entry : ""});
 }
 
+class ThreadLocalLogBufferManager : public Continuation
+{
+public:
+  ThreadLocalLogBufferManager()
+  {
+    this->thread_affinity = this_ethread();
+    SET_HANDLER(&ThreadLocalLogBufferManager::wakeup);
+
+    int period = Log::config->max_secs_per_buffer >> 1;
+    if (period < 1) {
+      period = 1;
+    }
+    eventProcessor.schedule_every(this, period * HRTIME_SECOND, ET_CALL);
+    Debug("log-config", "thread local buffer manager init: %d wakeup period", period);
+  }
+  ~ThreadLocalLogBufferManager() override
+  {
+    Debug("log-config", "thread local buffer manager destructor");
+    // only the LogBuffer objects are owned by this
+    for (auto [o, b] : current_buffers) {
+      // ideally we flush these here but there are shutdown order issues so if the
+      // logbuffer still exists at this point we have to drop it
+      delete b;
+    }
+    current_buffers.clear();
+  }
+  int
+  wakeup(int /* event ATS_UNUSED */, Event * /* e ATS_UNUSED */)
+  {
+    for (auto &[o, b] : current_buffers) {
+      if (b && ink_hrtime_to_sec(Thread::get_hrtime()) > b->expiration_time()) {
+        o->flush_buffer(b);
+        b = nullptr;
+      }
+    }
+
+    return EVENT_CONT;
+  }
+
+  LogBuffer *
+  current_buffer(LogObject *o, size_t *offset, size_t bytes_needed)
+  {
+    LogBuffer *buffer = current_buffers[o];
+    if (buffer == nullptr) {
+      buffer             = new LogBuffer(Log::config, o, Log::config->log_buffer_size);
+      current_buffers[o] = buffer;
+    }
+    if (buffer->add_entry(offset, bytes_needed) != LogBuffer::LB_OK) {
+      o->flush_buffer(buffer);
+
+      buffer             = new LogBuffer(Log::config, o, Log::config->log_buffer_size);
+      current_buffers[o] = buffer;
+      if (buffer->add_entry(offset, bytes_needed) != LogBuffer::LB_OK) {
+        return nullptr;
+      }
+    };
+    return buffer;
+  }
+
+private:
+  std::map<LogObject *, LogBuffer *> current_buffers;
+};
+
+/*
+ * This will return a LogBuffer object that is per-LogObject and per-Thread.  This function will handle all of
+ * the details around flushing buffers to the preproc threads and periodically checking for idle buffers.
+ */
+LogBuffer *
+thread_local_buffer(LogObject *o, size_t *offset, size_t bytes_needed)
+{
+  thread_local ThreadLocalLogBufferManager manager;
+  return manager.current_buffer(o, offset, bytes_needed);
+}
```

Comment:
I'm unapologetically anal about narrow scoping, so I suggest making this a public static member of ThreadLocalBufferManager, so that all other members can be private.